code/modules/wiremod/components/action/mmi.dm

```diff
@@ -0,0 +1,182 @@
+	display_name = "Man-Machine Interface"
+	display_desc = "A component that allows MMI to enter shells to send output signals."
+
+	/// The message to send to the MMI in the shell.
+	var/datum/port/input/message
+	/// Sends the current MMI a message
+	var/datum/port/input/send
+	/// Ejects the current MMI
+	var/datum/port/input/eject
+
+	/// Called when the MMI tries moving north
+	var/datum/port/output/north
+	/// Called when the MMI tries moving east
+	var/datum/port/output/east
+	/// Called when the MMI tries moving south
+	var/datum/port/output/south
+	/// Called when the MMI tries moving west
+	var/datum/port/output/west
+
+	/// Returns what the MMI last clicked on.
+	var/datum/port/output/clicked_atom
+	/// Called when the MMI clicks.
+	var/datum/port/output/attack
+	/// Called when the MMI right clicks.
+	var/datum/port/output/secondary_attack
+
+	/// The current MMI card
+	var/obj/item/mmi/brain
+
+	/// Maximum length of the message that can be sent to the MMI
+	var/max_length = 300
+
+/obj/item/circuit_component/mmi/Initialize()
+	. = ..()
+	message = add_input_port("Message", PORT_TYPE_STRING)
+	send = add_input_port("Send Message", PORT_TYPE_SIGNAL)
+	eject = add_input_port("Eject", PORT_TYPE_SIGNAL)
+
+	north = add_output_port("North", PORT_TYPE_SIGNAL)
+	east = add_output_port("East", PORT_TYPE_SIGNAL)
+	south = add_output_port("South", PORT_TYPE_SIGNAL)
+	west = add_output_port("West", PORT_TYPE_SIGNAL)
+
+	attack = add_output_port("Attack", PORT_TYPE_SIGNAL)
+	secondary_attack = add_output_port("Secondary Attack", PORT_TYPE_SIGNAL)
+	clicked_atom = add_output_port("Target Entity", PORT_TYPE_ATOM)
+
+/obj/item/circuit_component/mmi/Destroy()
+	remove_current_brain()
+	message = null
+	send = null
+	eject = null
+	north = null
+	east = null
+	south = null
+	west = null
+	attack = null
+	secondary_attack = null
+	clicked_atom = null
+	return ..()
+
+/obj/item/circuit_component/mmi/input_received(datum/port/input/port)
+	. = ..()
+	if(.)
+		return
+
+	if(!brain)
+		return
+
+	if(COMPONENT_TRIGGERED_BY(eject, port))
+		remove_current_brain()
+	if(COMPONENT_TRIGGERED_BY(send, port))
+		if(!message.input_value)
+			return
+
+		var/msg_str = copytext(html_encode(message.input_value), 1, max_length)
+
+		var/mob/living/target = brain.brainmob
+		if(!target)
+			return
+
+		to_chat(target, "[span_bold("You hear a message in your ear: ")][msg_str]")
+
+
+/obj/item/circuit_component/mmi/register_shell(atom/movable/shell)
+	. = ..()
+	RegisterSignal(shell, COMSIG_PARENT_ATTACKBY, .proc/handle_attack_by)
+
+/obj/item/circuit_component/mmi/unregister_shell(atom/movable/shell)
+	UnregisterSignal(shell, COMSIG_PARENT_ATTACKBY)
+	remove_current_brain()
+	return ..()
+
+/obj/item/circuit_component/mmi/proc/handle_attack_by(atom/movable/shell, obj/item/item, mob/living/attacker)
+	SIGNAL_HANDLER
+	if(istype(item, /obj/item/mmi))
+		var/obj/item/mmi/target_mmi = item
+		if(!target_mmi.brainmob)
+			return
+		add_mmi(item)
+
```

Comment:
Wouldn't you want a `COMPONENT_NO_AFTERATTACK` to stop the attack chain after this?